### PR TITLE
Make ui.Canvas.getDestinationClipBounds works with matrix

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/108154

cc @flar 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by flar — COMMENTED** [review]
Clicking "submit" on the 2 changes I've identified so I don't forget, but I'm still looking at the unit tests when I can find the chance.

File: display_list/display_list_builder.h [hunk 1]
```diff
@@ -211,7 +211,7 @@ class DisplayListBuilder final : public virtual Dispatcher,
   void clipRect(const SkRect& rect, SkClipOp clip_op, bool is_aa) override;
   void clipRRect(const SkRRect& rrect, SkClipOp clip_op, bool is_aa) override;
   void clipPath(const SkPath& path, SkClipOp clip_op, bool is_aa) override;
-
+  void intersect(const SkRect& rect);
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply display_list/display_list_builder.h]: Done

File: display_list/display_list_builder.cc [hunk 2]
```diff
@@ -653,15 +649,19 @@ void DisplayListBuilder::clipPath(const SkPath& path,
   switch (clip_op) {
     case SkClipOp::kIntersect:
       Push<ClipIntersectPathOp>(0, 1, path, is_aa);
-      if (!current_layer_->clip_bounds.intersect(path.getBounds())) {
-        current_layer_->clip_bounds.setEmpty();
-      }
+      intersect(path.getBounds());
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply display_list/display_list_builder.cc]: Done, I also added the logic for `inverse fill type` with `SkClipOp::kDifference`.

File: display_list/display_list_builder.cc [hunk 3]
```diff
@@ -653,15 +649,24 @@ void DisplayListBuilder::clipPath(const SkPath& path,
...
       Push<ClipIntersectPathOp>(0, 1, path, is_aa);
-      if (!current_layer_->clip_bounds.intersect(path.getBounds())) {
-        current_layer_->clip_bounds.setEmpty();
+      if (!path.isInverseFillType()) {
+        intersect(path.getBounds());
       }
       break;
     case SkClipOp::kDifference:
       Push<ClipDifferencePathOp>(0, 1, path, is_aa);
+      if (path.isInverseFillType()) {
```
Comment (flar) [inline display_list/display_list_builder.cc:658]: Delete this. kDifference is not an intersection. You can't call intersect here regardless of path type.

**Review by flar — CHANGES_REQUESTED** [review]

File: display_list/display_list_builder.cc (hunk 3 above)
Comment (flar) [reply display_list/display_list_builder.cc:658]: It might be possible to do a difference computation, but it can't be done on the bounds of the clip being "differenced" and it would involve a lot of round-corner-checking for rrect and path analysis for path. And after all of that you'd have to be really really really sure that it was guaranteed to lop off an entire side of the former clip bounds before you could commit to a change in the conservative estimate of the clip bounds. It's doable for some cases, but I didn't feel it was worth it to work through all of the corner conditions.

File: display_list/display_list_unittests.cc [hunk 4]
```diff
@@ -2252,6 +2314,34 @@ TEST(DisplayList, DiffClipPathDoesNotAffectClipBounds) {
...
 }
 
+TEST(DisplayList, ClipPathWithInvertFillTypeDoesNotAffectClipBounds) {
+  SkRect cull_rect = SkRect::MakeLTRB(0, 0, 100.0, 100.0);
+  DisplayListBuilder builder(cull_rect);
+  SkPath clip = SkPath().addCircle(10.2, 11.3, 2).addCircle(20.4, 25.7, 2);
+  clip.setFillType(SkPathFillType::kInverseWinding);
+  builder.clipPath(clip, SkClipOp::kIntersect, false);
+
+  SkRect initial_local_bounds = builder.getLocalClipBounds();
```
Comment (flar) [inline display_list/display_list_unittests.cc:2324]: "initial_"? That modifier would make sense if you were going to take measurements later on under a different name. It might be reasonable to just inline the getBounds calls into the ASSERT macros here.

File: display_list/display_list_unittests.cc [hunk 5]
```diff
@@ -2252,6 +2314,34 @@ TEST(DisplayList, DiffClipPathDoesNotAffectClipBounds) {
...
+  SkRect initial_destination_bounds = builder.getDestinationClipBounds();
+  ASSERT_EQ(initial_local_bounds, cull_rect);
+  ASSERT_EQ(initial_destination_bounds, cull_rect);
+}
+
+TEST(DisplayList, DiffClipPathWithInvertFillTypeAffectsClipBounds) {
+  SkRect cull_rect = SkRect::MakeLTRB(0, 0, 100.0, 100.0);
```
Comment (flar) [inline display_list/display_list_unittests.cc:2331]: Diff clip should not modify bounds even if the path is not inverted...

File: display_list/display_list_unittests.cc [hunk 6]
```diff
@@ -2116,6 +2116,25 @@ TEST(DisplayList, ClipRectAffectsClipBounds) {
...
 
+TEST(DisplayList, ClipRectAffectsClipBoundsWithMatrix) {
+  DisplayListBuilder builder;
+  SkRect clipBounds1 = SkRect::MakeLTRB(0, 0, 10, 10);
+  SkRect clipBounds2 = SkRect::MakeLTRB(10, 10, 20, 20);
+  builder.save();
+  builder.clipRect(clipBounds1, SkClipOp::kIntersect, false);
+  builder.translate(10, 0);
+  builder.clipRect(clipBounds1, SkClipOp::kIntersect, false);
+  ASSERT_EQ(builder.getDestinationClipBounds(), SkRect::MakeEmpty());
```
Comment (flar) [inline display_list/display_list_unittests.cc:2127]: Perhaps we should relax these to "get...Bounds().isEmpty()" rather than force a specific empty return value? Are they documented to return exactly that rectangle for an empty condition? I suppose we could just modify the test if we ever return a different value, but an implementation might decide to just max/min the 8 values and let their natural "emptiness" be a return value. We only get a hard-coded 0,0,0,0 rect because we use Skia for the intersections and they decided to return a false boolean with no changes to the rectangle for an empty intersection and so we have to manually force the answer to an empty answer (using MakeEmpty() for convenience).

File: testing/dart/canvas_test.dart [hunk 7]
```diff
@@ -731,6 +731,27 @@ void main() {
...
+    final PictureRecorder recorder = PictureRecorder();
+    final Canvas canvas = Canvas(recorder);
+    const Rect clipBounds1 = Rect.fromLTRB(0.0, 0.0, 10.0, 10.0);
+    const Rect clipBounds2 = Rect.fromLTRB(10.0, 10.0, 20.0, 20.0);
+
+    canvas.save();
+    canvas.clipRect(clipBounds1);
+    canvas.translate(0, 10.0);
+    canvas.clipRect(clipBounds1);
+    expect(canvas.getDestinationClipBounds(), Rect.fromLTRB(0.0, 0.0, 0.0, 0.0));
```
Comment (flar) [inline testing/dart/canvas_test.dart:744]: Same comment about measuring expectations against the "isEmpty" property rather than requiring a specific answer as I made above for the native tests.

File: display_list/display_list_unittests.cc (hunk 5 above)
Comment (flar) [reply display_list/display_list_unittests.cc:2331]: (Or ... we don't yet have the ability to compute its impact on the clip bounds, but that isn't something we should test...)

If anything, an inverse path with kDifference is probably like a normal path with kIntersect???

File: display_list/display_list_builder.cc (hunk 3 above)
Comment (flar) [reply display_list/display_list_builder.cc:658]: Hang on, I think I understand what is going on here. You are mapping "kDifference of inverse path" to "kIntersect of the original path".  Hmmm... I think that makes sense now.

Comment (flar) [reply display_list/display_list_builder.cc:658]: (I deleted a few comments down below now.)

File: display_list/display_list_unittests.cc (hunk 4 above)
Comment (ColdPaleLight) [reply display_list/display_list_unittests.cc]: Done.

File: display_list/display_list_unittests.cc (hunk 6 above)
Comment (ColdPaleLight) [reply display_list/display_list_unittests.cc]: Done.

File: testing/dart/canvas_test.dart (hunk 7 above)
Comment (ColdPaleLight) [reply testing/dart/canvas_test.dart]: Done.

File: display_list/display_list_builder.cc (hunk 3 above)
Comment (flar) [inline display_list/display_list_builder.cc:658]: Add a reminder comment here about how this works. Nothing too detailed, just reminding the reader (i.e. me?) that inverted difference is like intersection...

**Review by flar — APPROVED** [review]
Just a request to add a comment, but LGTM.

File: display_list/display_list_builder.cc (hunk 3 above)
Comment (ColdPaleLight) [reply display_list/display_list_builder.cc:659]: Done.
